pbj-core/pbj-grpc-helidon/src/main/java/com/hedera/pbj/grpc/helidon/PbjProtocolHandler.java

```diff
@@ -0,0 +1,590 @@
+     */
+    private ScheduledFuture<?> scheduleDeadline(@NonNull final String timeout) {
+        final var matcher = GRPC_TIMEOUT_PATTERN.matcher(timeout);
+        if (matcher.matches()) {
+            final var num = Integer.parseInt(matcher.group(1));
+            final var unit = matcher.group(2);
+            final var deadline = System.nanoTime() * TimeUnit.NANOSECONDS.convert(num, switch (unit) {
+                case "H" -> TimeUnit.HOURS;
+                case "M" -> TimeUnit.MINUTES;
+                case "S" -> TimeUnit.SECONDS;
+                case "m" -> TimeUnit.MILLISECONDS;
+                case "u" -> TimeUnit.MICROSECONDS;
+                case "n" -> TimeUnit.NANOSECONDS;
+                // This should NEVER be reachable, because the matcher would not have matched.
+                default -> throw new GrpcException(GrpcStatus.INTERNAL, "Invalid unit: " + unit);
+            });
+            return deadlineDetector.scheduleDeadline(deadline, () -> {
+                route.deadlineExceededCounter().increment();
+                incoming.onError(new GrpcException(GrpcStatus.DEADLINE_EXCEEDED));
+            });
+        }
+
+        return new NoopScheduledFuture();
+    }
+
+    /**
+     * According to the specification, the "Response-Headers" are:
+     *
+     * <pre>HTTP-Status [Message-Encoding] [Message-Accept-Encoding] Content-Type *Custom-Metadata</pre>
+     *
+     * <p>Where "Status" is <strong>always</strong> 200 OK.</p>
+     *
+     * <p>The Response-Headers are normally sent, followed by the data, followed by trailers. But if the request fails
+     * right away, before any handling, then it is possible to send Trailers-Only instead.
+     */
+    private void sendResponseHeaders(
+            @Nullable final Header messageEncoding,
+            @NonNull final HttpMediaType contentType,
+            @NonNull final List<Header> customMetadata) {
+
+        // Some headers are http2 specific, the rest are used for the grpc protocol
+        final var grpcHeaders = WritableHeaders.create();
+        final var http2Headers = Http2Headers.create(grpcHeaders);
+        // FUTURE: I think to support custom headers in the response, we would have to list them here.
+        // Since this has to be sent before we have any data to send, we must know ahead of time which
+        // custom headers are to be returned.
+        grpcHeaders.set(HeaderNames.TRAILER, "grpc-status, grpc-message");
+        http2Headers.status(Status.OK_200);
+        grpcHeaders.contentType(contentType);
+        grpcHeaders.set(GRPC_ACCEPT_ENCODING, IDENTITY);
+        customMetadata.forEach(grpcHeaders::set);
+        if (messageEncoding != null) {
+            grpcHeaders.set(messageEncoding);
+        }
+
+        streamWriter.writeHeaders(http2Headers,
+                streamId,
+                Http2Flag.HeaderFlags.create(Http2Flag.END_OF_HEADERS),
+                flowControl.outbound());
```

Comment:
Kind of seconding to Austin above, are the http2Headers connected to the grpcHeaders at line 389, so that whatever you set at line 399 is available in the http2Headers? Because you never use the grpcHeaders directly after line 399.

Reply:
@rbair23 I've reordered how these headers are set, so that the GRPC headers are all set, and then copied into http headers.

I'm assuming for now that the previous ordering was just a mistake, so let me know if there was some intention here that I'm not understanding.

Reply:
The updated code looks good to me now.